include/picolibrary/testing/interactive/spi.h

```diff
@@ -23,10 +23,96 @@
+#include <limits>
+#include <utility>
+
+#include "picolibrary/format.h"
+
 /**
  * \brief Serial Peripheral Interface (SPI) interactive testing facilities.
  */
 namespace picolibrary::Testing::Interactive::SPI {
+
+/**
+ * \brief SPI controller echo interactive test helper.
+ *
+ * \tparam Output_Stream The type of asynchronous serial output stream to use to output
+ *         information to the user.
+ * \tparam Transmitter The type of asynchronous serial transmitter to use to transmit
+ *         information to the user.
+ * \tparam Controller The type of SPI controller to test.
+ * \tparam Delayer A nullary functor called to introduce a delay each time data is
+ *         exchanged.
+ *
+ * \param[in] transmitter The asynchronous serial transmitter to use to transmit
+ *            information to the user.
+ * \param[in] controller The SPI controller to test.
+ * \param[in] configuration The SPI controller clock configuration to use for the test.
+ * \param[in] delay The nullary functor called to introduce a delay each time data is
+ *            exchanged.
+ */
+template<template<typename> typename Output_Stream, typename Transmitter, typename Controller, typename Delayer>
+void echo( Transmitter transmitter, Controller controller, typename Controller::Configuration configuration, Delayer delay ) noexcept
+{
+    // #lizard forgives the length
+
+    auto stream = Output_Stream{ std::move( transmitter ) };
+
+    if ( stream.initialize().is_error() ) {
+        return;
+    } // if
+
+    {
+        auto const result = controller.initialize();
+        if ( result.is_error() ) {
+            static_cast<void>(
+                stream.print( "controller initialization error: {}\n", result.error() ) );
+
+            return;
+        } // if
+    }
+
+    {
+        auto const result = controller.configure( configuration );
+        if ( result.is_error() ) {
+            static_cast<void>(
+                stream.print( "controller configuration error: {}\n", result.error() ) );
+
+            return;
+        } // if
+    }
+
+    for ( ;; ) {
+        auto value = std::numeric_limits<std::uint8_t>::min();
+        do {
+            delay();
+
+            auto const result = controller.exchange( value );
+            if ( result.is_error() ) {
+                static_cast<void>( stream.print( "data exchange error: {}\n", result.error() ) );
+
+                return;
+            } // if
+
+            if ( stream
+                     .print(
+                         "exchange( {} ) -> {}\n",
+                         Format::Hexadecimal{ value },
+                         Format::Hexadecimal{ result.value() } )
+                     .is_error() ) {
+                return;
+            } // if
+
+            if ( result.value() != value ) {
+                static_cast<void>( stream.put( "echo error\n" ) );
+
+                return;
+            } // if
+        } while ( value < std::numeric_limits<std::uint8_t>::max() );
```

Comment:
`value` is never incremented and thus the do while loop will never terminate.

The do while loop is not actually necessary and can be replaced by having the enclosing for loop initialize and increment `value`:
```c++
for ( auto value = std::uint8_t{};; ++value ) {
} // for
```